File: backend_fastapi/core/storage.py
from datetime import datetime, timezone
from typing import Dict, Optional
from uuid import uuid4

import bcrypt

from core.demo_scenarios import DEMO_SCENARIOS
from models import AnalysisResult, StoredUser, UserProfile


class StorageManager:
    def __init__(self) -> None:
        self._analyses: Dict[str, AnalysisResult] = {}
        self._users: Dict[str, StoredUser] = {}
        self._seed_demo_data()

    def _seed_demo_data(self) -> None:
        for scenario in DEMO_SCENARIOS:
            if scenario.resultData and scenario.resultData.id:
                self._analyses[scenario.resultData.id] = scenario.resultData

        demo_email = 'ali.khan@example.com'
        if demo_email not in self._users:
            self._users[demo_email] = StoredUser(
                id='user-demo-ali',
                name='Ali Khan',
                email=demo_email,
                password=bcrypt.hashpw('demopassword123'.encode('utf-8'), bcrypt.gensalt()).decode('utf-8'),
                createdAt=datetime.now(timezone.utc).isoformat(),
            )

    def save_analysis(self, analysis: AnalysisResult) -> AnalysisResult:
        if not analysis.id:
            analysis.id = f'analysis-{int(datetime.now(timezone.utc).timestamp() * 1000)}-{uuid4().hex[:5]}'
        if not analysis.createdAt:
            analysis.createdAt = datetime.now(timezone.utc).isoformat()
        self._analyses[analysis.id] = analysis
        return analysis

    def get_analysis_by_id(self, analysis_id: str) -> Optional[AnalysisResult]:
        return self._analyses.get(analysis_id)

    def get_all_analyses(self) -> list[AnalysisResult]:
        return sorted(
            self._analyses.values(),
            key=lambda a: datetime.fromisoformat(a.createdAt),
            reverse=True,
        )

    def delete_analysis(self, analysis_id: str) -> bool:
        return analysis_id in self._analyses and bool(self._analyses.pop(analysis_id))

    def create_user(self, name: str, email: str, password: Optional[str] = None) -> UserProfile:
        lower_email = email.lower().strip()
        user = StoredUser(
            id=f'user-{int(datetime.now(timezone.utc).timestamp() * 1000)}-{uuid4().hex[:5]}',
            name=name or 'User',
            email=lower_email,
            password=bcrypt.hashpw(password.encode('utf-8'), bcrypt.gensalt()).decode('utf-8') if password else None,
            createdAt=datetime.now(timezone.utc).isoformat(),
        )
        self._users[lower_email] = user
        return UserProfile(id=user.id, name=user.name, email=user.email)

    def verify_password(self, plain: str, hashed: str) -> bool:
        if not hashed:
            return False
        try:
            return bcrypt.checkpw(plain.encode('utf-8'), hashed.encode('utf-8'))
        except Exception:
            return False

    def get_user_by_email(self, email: str) -> Optional[StoredUser]:
        return self._users.get(email.lower().strip())


storage = StorageManager()

File: src/server/storage.ts
import bcrypt from 'bcryptjs';
import { AnalysisResult, UserProfile } from '../types';
import { DEMO_SCENARIOS } from '../data/demoScenarios';

export interface StoredUser {
  id: string;
  name: string;
  email: string;
  password?: string;
  createdAt: string;
}

export class StorageManager {
  private analyses: Map<string, AnalysisResult> = new Map();
  private users: Map<string, StoredUser> = new Map();

  constructor() {
    this.seedDemoData();
  }

  public seedDemoData(): void {
    // Seed demo scenarios analyses
    for (const scenario of DEMO_SCENARIOS) {
      if (scenario.resultData && scenario.resultData.id) {
        this.analyses.set(scenario.resultData.id, scenario.resultData);
      }
    }

    // Seed default demo user
    const demoEmail = 'ali.khan@example.com';
    if (!this.users.has(demoEmail)) {
      const salt = bcrypt.genSaltSync(10);
      const hash = bcrypt.hashSync('demopassword123', salt);
      this.users.set(demoEmail, {
        id: 'user-demo-ali',
        name: 'Ali Khan',
        email: demoEmail,
        password: hash,
        createdAt: new Date().toISOString(),
      });
    }
  }

  public saveAnalysis(analysis: AnalysisResult): AnalysisResult {
    const id = analysis.id || `analysis-${Date.now()}-${Math.random().toString(36).substring(2, 7)}`;
    analysis.id = id;
    if (!analysis.createdAt) {
      analysis.createdAt = new Date().toISOString();
    }
    this.analyses.set(id, analysis);
    return analysis;
  }

  public getAnalysisById(id: string): AnalysisResult | null {
    return this.analyses.get(id) || null;
  }

  public getAllAnalyses(): AnalysisResult[] {
    const list = Array.from(this.analyses.values());
    list.sort((a, b) => new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime());
    return list;
  }

  public deleteAnalysis(id: string): boolean {
    return this.analyses.delete(id);
  }

  public createUser(name: string, email: string, password?: string): UserProfile {
    const id = `user-${Date.now()}-${Math.random().toString(36).substring(2, 7)}`;
    const lowerEmail = email.toLowerCase();
    let hashedPassword = '';
    if (password) {
      const salt = bcrypt.genSaltSync(10);
      hashedPassword = bcrypt.hashSync(password, salt);
    }

    const user: StoredUser = {
      id,
      name: name || 'User',
      email: lowerEmail,
      password: hashedPassword,
      createdAt: new Date().toISOString(),
    };

    this.users.set(lowerEmail, user);
    return {
      id: user.id,
      name: user.name,
      email: user.email,
    };
  }

  public verifyPassword(plain: string, hashed: string): boolean {
    if (!hashed) return false;
    try {
      return bcrypt.compareSync(plain, hashed);
    } catch {
      return false;
    }
  }

  public getUserByEmail(email: string): StoredUser | null {
    return this.users.get(email.toLowerCase()) || null;
  }
}

export const storage = new StorageManager();
